### PR TITLE
[Generic Database Connector] Make validate_configuration a top-level function and use it for composition

### PR DESCRIPTION
**Pull request by timgrein**
## Related to https://github.com/elastic/enterprise-search-team/issues/4067

**\~ 80% of LOC is due to parametrized testing**

- Created a new function `validate_non_empty_config_fields`, which can be used by arbitrary sources to validate, that mandatory configuration fields are non-empty
- Created a new function `validate_db_connection_fields`, which specifies validation for common database connection fields (`host`, `port`, `user`, `password`)
- Remove the datasource specific checks from the `generic_database` module and implement them at the right places in `oracle.py` and `postgresql.py`
- Mark two places with TODOs, which are currently depending on open PRs (will be removed by me as soon as the PRs are merged)
- Add error classes specific for the validation of a configuration:
  - `ConfigurationFieldMissingError`
  - `ConfigurationFieldEmptyError`
  - `ConfigurationFieldTypeError`
  - `ConfigurationValueError`

#### Pre-Review Checklist
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
~- [ ] Considered corresponding documentation changes~
~- [ ] Contributed any configuration settings changes to the configuration reference~

### CODE REVIEW

**Review by seanstory — APPROVED** [review]
great tests, as always. 👏

File: connectors/source.py [hunk 1]
```diff
@@ -20,6 +20,31 @@
...
+
+class ConfigurationFieldTypeError(ValueError):
+    pass
+
+
+class ConfigurationValueError(ValueError):
+    pass
+
+
+def validate_non_empty_config_fields(expected_fields, configuration):
```
Comment (tarekziade) [inline connectors/source.py:39]: that's a very long name

**Review by tarekziade — COMMENTED** [review]
Looks good but I feel like we're doing a lot of code where we could be more declarative and generic.

One idea: the configuration is defined by this function here 

https://github.com/elastic/connectors-python/blob/main/connectors/sources/generic_database.py#L67-L114

Maybe could add a new flag there to facilitate validation.

so host for instance, becomes:

```python
"host": {
  "value": "127.0.0.1",
  "label": "Host",
  "type": "str",
  "empty_allowed:": False
},
```

if you implement it this we, we can avoid adding new code for each specific validation sets (databases, etc)

File: connectors/source.py [hunk 2]
```diff
@@ -50,7 +82,12 @@ def _convert(self, value, type_):
         elif type_ == "bool":
             return value.lower() in ("y", "yes", "true", "1")
         elif type_ == "list":
-            return [item.strip() for item in value.split(",")]
+            return list(
```
Comment (tarekziade) [inline connectors/source.py:85]: you can simplify this 

```
return [item.strip() for item in value.split(",") if len(item.strip()) > 0]
```

**Comment by timgrein** [comment]
@tarekziade Thank you for the list comprehension suggestions, added them!

> One idea: the configuration is defined by this function here
> https://github.com/elastic/connectors-python/blob/main/connectors/sources/generic_database.py#L67-L114
> Maybe could add a new flag there to facilitate validation.

Makes sense, we could translate the configuration to a schema, which is understood by `fastjsonschema`. I would prefer to merge this now as is and check first, what's the scope and the anticipated changes of rich-configurable fields for the python backend. I'll create an issue for it!

File: connectors/sources/generic_database.py [hunk 3]
```diff
@@ -36,6 +40,33 @@ def table_filter(table):
...
+    db_server_connection_fields = ["host", "port", "user", "password"]
+
+    validate_non_empty_config_fields(db_server_connection_fields, configuration)
+
+    port = configuration["port"]
+
+    if isinstance(port, str) and not port.isnumeric():
+        raise ConfigurationFieldTypeError(
+            f"Configured port {port} has to be an integer."
+        )
```
Comment (wangch079) [inline connectors/sources/generic_database.py:67]: I feel all these can be moved to `basic_validation`.

**Comment by artem-shelkovnikov** [comment]
> Looks good but I feel like we're doing a lot of code where we could be more declarative and generic.
> 
> One idea: the configuration is defined by this function here
> 
> https://github.com/elastic/connectors-python/blob/main/connectors/sources/generic_database.py#L67-L114
> 
> Maybe could add a new flag there to facilitate validation.

That's IMO tightly coupled with Rich Configuration fields work - some of these validations that can be declarative will have to be properly reflected in the frontend - such as "non-empty" for instance. It'll be weird to validate in on a backend only IMO.

File: connectors/source.py [hunk 4]
```diff
@@ -92,6 +124,14 @@ def get(self, key, default=None):
...
         return name in self._config
 
+    def any_field_is_empty(self, names):
+        if any(not self.has_field(name) for name in names):
+            raise ConfigurationFieldMissingError
+
+        empty_fields = [name for name in names if self._config[name].is_empty()]
+
+        return len(empty_fields) > 0, empty_fields
+
```
Comment (artem-shelkovnikov) [inline connectors/source.py:134]: This can probably be a separate PR with separate test

File: connectors/source.py [hunk 5]
```diff
@@ -92,6 +124,14 @@ def get(self, key, default=None):
     def has_field(self, name):
         return name in self._config
 
+    def any_field_is_empty(self, names):
+        if any(not self.has_field(name) for name in names):
+            raise ConfigurationFieldMissingError
```
Comment (artem-shelkovnikov) [inline connectors/source.py:129]: Should this error mention which field is missing? Ideally it should include all fields that are missing IMO

File: connectors/source.py [hunk 6]
```diff
@@ -20,6 +20,31 @@
 )
 
 
+class ConfigurationFieldMissingError(ValueError):
```
Comment (artem-shelkovnikov) [inline connectors/source.py:23]: Why different exception types?

I don't see any code catching these types, so it might be an overkill?

File: connectors/source.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (timgrein) [reply connectors/source.py:39]: I'm already checking that in the corresponding project brief, if that's also in the scope.

File: connectors/sources/generic_database.py (hunk 3 above)
Comment (timgrein) [reply connectors/sources/generic_database.py:67]: This is specific to databases, so I don't think it's generic enough to have it in a `basic_validation` method inside `DataSourceConfiguration`.

File: connectors/source.py (hunk 6 above)
Comment (timgrein) [reply connectors/source.py:23]: I thought it's useful to keep them distinct from each other, especially if we provide specific constructors, e.g. as you've pointed out we can pass the missing fields to `ConfigurationFieldMissingError` and for `ConfigurationFieldTypeError ` we could pass the expected type, actual type, field and value.

File: connectors/source.py (hunk 4 above)
Comment (timgrein) [reply connectors/source.py:134]: It's tested in `test_any_field_is_empty` and two other tests in `test_source.py`. I can create a separate PR for the new error classes and this functionality 👍 Just don't want to overload our open PRs. We already have ~ 15 PRs open since one week.

Comment (artem-shelkovnikov) [reply connectors/source.py:134]: Yeah makes sense, just a note for the future :) I love small PRs and take much less time approving them :)

File: connectors/source.py (hunk 6 above)
Comment (artem-shelkovnikov) [reply connectors/source.py:23]: Yeah makes sense, I think we can extend these error classes later with specific constructors, awesome!

File: connectors/source.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (navarone-feekery) [reply connectors/source.py:39]: Mandatory fields are indeed in the Rich Configurable Fields scope. Empty validation will rely on mandatory fields and also take defaults (if any) into consideration.

[the remark this answers is not in the document]
Comment (navarone-feekery) [reply connectors/source.py:39]: My general thoughts are; if you need this validation to fulfill some other development requirements, it's okay if you add it. I will just build out from what you've added or refactor it if it doesn't fit the implementation goals of Rich Configurable Fields.
Unfortunately I haven't written an implementation doc yet so I can't say whether the approach would be the same as yours here.

**Comment by timgrein** [comment]
Closed as it has too much overlap with rich-configurable fields.
